artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/PagedReferenceImpl.java

```diff
@@ -50,14 +47,24 @@ public class PagedReferenceImpl extends LinkedListImpl.Node<PagedReferenceImpl>
 
    private int messageEstimate = -1;
 
+   PagePosition position;
+
+   @Override
+   public PagePosition getPosition() {
+      if (position == null) {
+         position = getPagedMessage().newPositionObject();
+      }
+      return position;
+   }
```

Comment:
Can this be called from different threads? There are synchronized and volatile uses elsewhere in the class (including the method it calls) suggesting it could be. If so this would not seem safe.

Reply:
Instead of making getPosition() synchronized, I will just make positon volatile here.

This is used on browsing. And this is just caching the value.

Reply:
It may not be necessary for it to be volatile, or may still be insufficient that it is...depends what the implications are of two threads creating and using their own independent PagePosition objects for the same message, either because they either didnt see another threads update of the caching variable, or because two race to create it and both update the variable, the latter of which remains possible even if the field is volatile.

I'm not sure how its used overall, but peaking at what newPositionObject() does, it seems like it just created holding values it gets from the paged message (for which the getter method this calls to retrieve is synchronized). The real question would be around whether any of the various values the position object seems to hold are ever manipulated via the setters, as then having 'duplicate' objects could be problematic.